src/neml2/models/VariableStore.cxx

```diff
@@ -181,4 +289,84 @@ VariableStore::collect_output_second_derivatives() const
   return sec_derivs;
 }
 
+torch::jit::Stack
+VariableStore::collect_input_stack() const
+{
+  torch::jit::Stack stack;
+  const auto & vars = input_axis().variable_names();
+  stack.reserve(vars.size());
+  for (const auto & name : vars)
+    stack.push_back(input_variable(name).tensor());
+  return stack;
+}
+
+torch::jit::Stack
+VariableStore::collect_output_stack(bool out, bool dout, bool d2out) const
+{
+  neml_assert_dbg(out || dout || d2out,
+                  "At least one of the output/derivative flags must be true.");
+
+  const auto & yvars = output_axis().variable_names();
+  const auto & xvars = input_axis().variable_names();
+
+  std::vector<torch::Tensor> stacklist;
+  std::vector<std::uint8_t> sparsity;
+
+  if (out)
+  {
+    sparsity.insert(sparsity.end(), yvars.size(), 1);
+    for (const auto & yvar : yvars)
+      stacklist.push_back(output_variable(yvar).tensor());
+  }
+
+  if (dout)
+  {
+    for (const auto & yvar : yvars)
+    {
+      const auto & derivs = output_variable(yvar).derivatives();
+      for (const auto & xvar : xvars)
+      {
+        const auto & deriv = derivs.find(xvar);
```

Comment:
This initially bothered me, but I guess `find` is log(n) or something in the number of keys.  So no changes required, just making sure you thought about the cost (and similar for d2out).

Reply:
I am also not satisfied with the cost. The root cause of this is my poor choice of container for variable derivatives. I plan to fix that (along with every consequent mistake) in the future.